silo/api.py

```diff
@@ -302,28 +300,59 @@ def data(self, request, id):
         # calling get_object applies the permission classes to this query
         self.get_object()
 
+        # get the mongo collection
+        collection = LabelValueStore._get_collection()
+
         draw = int(request.GET.get("draw", 1))
         offset = int(request.GET.get('start', -1))
         length = int(request.GET.get('length', 10))
 
         # filtering syntax is the mongodb syntax
-        query = request.GET.get('query',"{}")
-        filter_fields = json.loads(query)
-
-        recordsTotal = LabelValueStore.objects(silo_id=id, **filter_fields).count()
-
-        # workaround until the problem of javascript not increasing the value of length is fixed
+        query = request.GET.get('query', '{}')
+        group = request.GET.get('group', '{}')
+        sort = str(request.GET.get('sort', '{}'))
+        query_fields = json.loads(query)
+        group_fields = json.loads(group)
+        sort_fields = json.loads(sort)
+
+        # creating the aggregation pipeline
+        pipeline = [
+            {'$match': {'$and': [{'silo_id': int(id)}, query_fields]}},
+            {'$project': {
+                'create_date': 0,
+                'edit_date': 0,
+                'silo_id': 0,
+                'read_id': 0
+            }}
+        ]
+        if group_fields:
+            pipeline.append({'$group': group_fields})
+        if sort_fields:
+            pipeline.append({'$sort': sort_fields})
+
+        count_pipeline = pipeline + [{'$count': 'count'}]
+        cmd_cur = collection.aggregate(pipeline=count_pipeline)
+        result = list(cmd_cur)[0]
+        records_total = result['count']
+
+        # workaround until the problem of javascript not increasing
+        # the value of length is fixed
         if offset >= 0:
             length = offset + length
-            data = LabelValueStore.objects(silo_id=id, **filter_fields).exclude('create_date', 'edit_date', 'silo_id','read_id').skip(offset).limit(length)
+            pipeline.append({'$skip': offset})
+            pipeline.append({'$limit': length})
+            cmd_cur = collection.aggregate(pipeline=pipeline)
+            result = list(cmd_cur)
+            data = JSONEncoder().encode(result)
         else:
-            data = LabelValueStore.objects(silo_id=id, **filter_fields).exclude('create_date', 'edit_date', 'silo_id','read_id')
-
-        sort = str(request.GET.get('sort',''))
-        data = data.order_by(sort)
-        json_data = json.loads(data.to_json())
-
-        return JsonResponse({"data": json_data, "draw": draw, "recordsTotal": recordsTotal, "recordsFiltered": recordsTotal}, safe=False)
+            cmd_cur = collection.aggregate(pipeline=pipeline)
```

Comment:
This should not be there. This operation was done before.

Reply:
The first operation only counts the number of rows. This one we return the data.